packages/instruction-plans/src/transaction-planner-base.ts

```diff
@@ -0,0 +1,416 @@
+    instructionPlan: SingleInstructionPlan,
+    context: TraverseContext,
+): Promise<MutableTransactionPlan | null> {
+    const predicate = (message: BaseTransactionMessage & TransactionMessageWithFeePayer) =>
+        appendTransactionMessageInstructions([instructionPlan.instruction], message);
+    const candidate = await selectAndMutateCandidate(context, context.parentCandidates, predicate);
+    if (candidate) {
+        return null;
+    }
+    const message = await createNewMessage(context, instructionPlan, predicate);
+    return { kind: 'single', message };
+}
+
+async function traverseMessagePacker(
+    instructionPlan: MessagePackerInstructionPlan,
+    context: TraverseContext,
+): Promise<MutableTransactionPlan | null> {
+    const messagePacker = instructionPlan.getMessagePacker();
+    const transactionPlans: SingleTransactionPlan[] = [];
+    const candidates = [...context.parentCandidates];
+
+    while (!messagePacker.done()) {
+        const candidate = await selectAndMutateCandidate(context, candidates, messagePacker.packMessageToCapacity);
+        if (!candidate) {
+            const message = await createNewMessage(context, instructionPlan, messagePacker.packMessageToCapacity);
+            const newPlan: MutableSingleTransactionPlan = { kind: 'single', message };
+            transactionPlans.push(newPlan);
+        }
+    }
+
+    if (transactionPlans.length === 1) {
+        return transactionPlans[0];
+    }
+    if (transactionPlans.length === 0) {
+        return null;
+    }
+    if (context.parent?.kind === 'parallel') {
+        return { kind: 'parallel', plans: transactionPlans };
+    }
+    return {
+        divisible: context.parent?.kind === 'sequential' ? context.parent.divisible : true,
+        kind: 'sequential',
+        plans: transactionPlans,
+    };
+}
+
+function getSequentialCandidate(latestPlan: MutableTransactionPlan): MutableSingleTransactionPlan | null {
+    if (latestPlan.kind === 'single') {
+        return latestPlan;
+    }
+    if (latestPlan.kind === 'sequential' && latestPlan.plans.length > 0) {
+        return getSequentialCandidate(latestPlan.plans[latestPlan.plans.length - 1]);
+    }
+    return null;
+}
+
+function getParallelCandidates(latestPlan: TransactionPlan): MutableSingleTransactionPlan[] {
+    return getAllSingleTransactionPlans(latestPlan);
+}
+
+async function selectAndMutateCandidate(
+    context: Pick<TraverseContext, 'abortSignal' | 'onTransactionMessageUpdated'>,
+    candidates: MutableSingleTransactionPlan[],
+    predicate: (
+        message: BaseTransactionMessage & TransactionMessageWithFeePayer,
+    ) => BaseTransactionMessage & TransactionMessageWithFeePayer,
+): Promise<MutableSingleTransactionPlan | null> {
+    for (const candidate of candidates) {
+        try {
+            const message = await context.onTransactionMessageUpdated(predicate(candidate.message), {
+                abortSignal: context.abortSignal,
+            });
```

Comment:
The user-supplied `onTransactionMessageUpdated` function could fail to make use of the `AbortSignal` here. I think we need to double check them. It's their funeral if they don't cancel _their_ work, but in any case we should cancel _ours_.
```suggestion
            const message = await getAbortablePromise(
                context.onTransactionMessageUpdated(predicate(candidate.message), {
                    abortSignal: context.abortSignal,
                },
                context.abortSignal,
            );
```

Reply:
Oh good to know thanks! I had to wrap the first argument in a `Promise.resolve` otherwise typescript was not happy since `getAbortablePromise` require a `Promise<T>`.